### PR TITLE
Arrow dictionary encoding

### PR DESCRIPTION
**Pull request by fexolm**

### CODE REVIEW

File: DataMgr/ForeignStorage/ArrowCsvForeignStorage.cpp [hunk 1]
```diff
@@ -241,36 +245,64 @@ void ArrowCsvForeignStorage::registerTable(std::pair<int, int> table_key,
     col.resize(fragments.size());
     auto clp = ARROW_GET_DATA(table.column(cln++)).get();
 
+    if (c.columnType.is_dict_encoded_string()) {
+      auto dictDesc = const_cast<DictDescriptor*>(
+          catalog->getMetadataForDict(c.columnType.get_comp_param()));
+      CHECK(dictDesc);
+      auto stringDict = dictDesc->stringDict.get();
+      g_dictionaries[col_key] = stringDict;
```
Comment (fexolm) [inline DataMgr/ForeignStorage/ArrowCsvForeignStorage.cpp:254]: we can get rid of that global dictionary and use a local variable instead

File: Catalog/Catalog.cpp [hunk 2]
```diff
@@ -2135,7 +2127,9 @@ void Catalog::createTable(
     removeTableFromMap(td.tableName, td.tableId, true);
     throw;
   }
-
+  if (!td.storageType.empty()) {
+    ForeignStorageInterface::registerTable(this, getCurrentDB().dbId, td, cds);
+  }
```
Comment (anton-malakhov) [inline Catalog/Catalog.cpp:2133]: at least, leave a comment that it breaks the transaction

File: Catalog/Catalog.cpp [hunk 3]
```diff
@@ -2395,9 +2389,9 @@ void Catalog::setColumnDictionary(ColumnDescriptor& cd,
                     folderPath,
                     false);
   dds.push_back(dd);
-  if (!cd.columnType.is_array()) {
-    cd.columnType.set_size(cd.columnType.get_comp_param() / 8);
-  }
+  // if (!cd.columnType.is_array()) {
+  //   cd.columnType.set_size(cd.columnType.get_comp_param() / 8);
+  // }
```
Comment (anton-malakhov) [inline Catalog/Catalog.cpp:2396]: bad, bad, bad :)

File: Catalog/Catalog.cpp [hunk 4]
```diff
@@ -2135,7 +2127,9 @@ void Catalog::createTable(
     removeTableFromMap(td.tableName, td.tableId, true);
     throw;
   }
-
+  if (!td.storageType.empty()) {
```
Comment (anton-malakhov) [inline Catalog/Catalog.cpp:2131]:
```suggestion
  // TODO: rollback if exception is raised
  if (!td.storageType.empty()) {
```

File: Catalog/Catalog.cpp [hunk 5]
```diff
@@ -2395,9 +2389,9 @@ void Catalog::setColumnDictionary(ColumnDescriptor& cd,
                     folderPath,
                     false);
   dds.push_back(dd);
-  if (!cd.columnType.is_array()) {
-    cd.columnType.set_size(cd.columnType.get_comp_param() / 8);
-  }
+  // if (!cd.columnType.is_array()) {
```
Comment (anton-malakhov) [inline Catalog/Catalog.cpp:2394]:
```suggestion
  // TODO: Why??? fix back
  // if (!cd.columnType.is_array()) {
```

File: DataMgr/ForeignStorage/ArrowCsvForeignStorage.cpp [hunk 6]
```diff
@@ -159,7 +162,8 @@ void ArrowCsvForeignStorage::prepareTable(const int db_id,
   td.hasDeletedCol = false;
 }
 
-void ArrowCsvForeignStorage::registerTable(std::pair<int, int> table_key,
+void ArrowCsvForeignStorage::registerTable(Catalog_Namespace::Catalog* catalog,
```
Comment (anton-malakhov) [inline DataMgr/ForeignStorage/ArrowCsvForeignStorage.cpp:166]:
```suggestion
// TODO: optimize the number of arguments
void ArrowCsvForeignStorage::registerTable(Catalog_Namespace::Catalog* catalog,
```

**Review by anton-malakhov — CHANGES_REQUESTED** [review]
please accept TODO comments

**Pull request merged**